File: sdk/typescript/rooch-sdk/src/keypairs/secp256k1/publickey.ts
// Copyright (c) RoochNetwork
// SPDX-License-Identifier: Apache-2.0

import { secp256k1 } from '@noble/curves/secp256k1'

import { AddressView } from '../../address/index.js'
import { PublicKey, PublicKeyInitData, SIGNATURE_SCHEME_TO_FLAG } from '../../crypto/index.js'
import { fromB64, sha256, toHEX } from '../../utils/index.js'

const SCHNORR_PUBLIC_KEY_SIZE = 32

/**
 * A Secp256k1 public key
 */
export class Secp256k1PublicKey extends PublicKey<AddressView> {
  static SIZE = SCHNORR_PUBLIC_KEY_SIZE

  private readonly data: Uint8Array

  /**
   * Create a new Secp256k1PublicKey object
   * @param value secp256k1 public key as buffer or base-64 encoded string
   */
  constructor(value: PublicKeyInitData) {
    super()

    if (typeof value === 'string') {
      this.data = fromB64(value)
    } else if (value instanceof Uint8Array) {
      this.data = value
    } else {
      this.data = Uint8Array.from(value)
    }

    if (this.data.length !== SCHNORR_PUBLIC_KEY_SIZE && this.data.length !== 33) {
      throw new Error(
        `Invalid public key input. Expected ${SCHNORR_PUBLIC_KEY_SIZE} bytes, got ${this.data.length}`,
      )
    }
  }

  /**
   * Checks if two Secp256k1 public keys are equal
   */
  // override equals(publicKey: Secp256k1PublicKey): boolean {
  //   return super.equals(publicKey)
  // }

  equals(publicKey: Secp256k1PublicKey): boolean {
    return super.equals(publicKey)
  }

  /**
   * Return the byte array representation of the Secp256k1 public key
   */
  override toBytes(): Uint8Array {
    return this.data
  }

  override toString(): string {
    return toHEX(this.data)
  }

  /**
   * Return the Bitcoin address associated with this Secp256k1 public key
   */
  override toAddress(): AddressView {
    return new AddressView(this.data)
  }

  /**
   * Return the Rooch address associated with this Secp256k1 public key
   */
  flag(): number {
    return SIGNATURE_SCHEME_TO_FLAG['Secp256k1']
  }

  /**
   * Verifies that the signature is valid for the provided message
   */
  async verify(message: Uint8Array, signature: Uint8Array): Promise<boolean> {
    return secp256k1.verify(
      secp256k1.Signature.fromCompact(signature),
      sha256(message),
      this.toBytes(),
    )
  }
}

File: sdk/golang/rooch-sdk/keypairs/secp256k1/publickey.go
package secp256k1

import (
	"encoding/base64"
	"encoding/hex"
	"errors"
	"github.com/decred/dcrd/dcrec/secp256k1/v4"
	"github.com/rooch-network/rooch/sdk/golang/address"
	"github.com/rooch-network/rooch/sdk/golang/crypto"
	"github.com/rooch-network/rooch/sdk/golang/utils"
)

const (
	SchnorrPublicKeySize = 32
)

// Secp256k1PublicKey represents a Secp256k1 public key
type Secp256k1PublicKey struct {
	data []byte
}

// NewSecp256k1PublicKey creates a new Secp256k1PublicKey object
func NewSecp256k1PublicKey(value interface{}) (*Secp256k1PublicKey, error) {
	var data []byte

	switch v := value.(type) {
	case string:
		// Assume base64 encoded string
		decoded, err := base64.StdEncoding.DecodeString(v)
		if err != nil {
			return nil, err
		}
		data = decoded
	case []byte:
		data = v
	default:
		return nil, errors.New("unsupported public key input type")
	}

	if len(data) != SchnorrPublicKeySize && len(data) != 33 {
		return nil, errors.New("invalid public key input size")
	}

	return &Secp256k1PublicKey{
		data: data,
	}, nil
}

// Equals checks if two Secp256k1 public keys are equal
func (pk *Secp256k1PublicKey) Equals(other *Secp256k1PublicKey) bool {
	return crypto.BytesEqual(pk.data, other.data)
}

// ToBytes returns the byte array representation of the Secp256k1 public key
func (pk *Secp256k1PublicKey) ToBytes() []byte {
	return pk.data
}

// String returns the hex string representation of the public key
func (pk *Secp256k1PublicKey) String() string {
	return hex.EncodeToString(pk.data)
}

// ToAddress returns the Bitcoin address associated with this Secp256k1 public key
func (pk *Secp256k1PublicKey) ToAddress() *address.AddressView {
	return address.NewAddressView(pk.data)
}

// ToAddressWith returns the Bitcoin address with specified network type
func (pk *Secp256k1PublicKey) ToAddressWith(network address.BitcoinNetworkType) *address.AddressView {
	return address.NewAddressViewWithNetwork(pk.data, network)
}

// Flag returns the signature scheme flag for Secp256k1
func (pk *Secp256k1PublicKey) Flag() int {
	return crypto.SignatureSchemeToFlag["Secp256k1"]
}

// Verify verifies that the signature is valid for the provided message
func (pk *Secp256k1PublicKey) Verify(message []byte, signature []byte) (bool, error) {
	// Create hash of the message
	messageHash := utils.Sha256(message)
	
	// Parse the signature
	sig, err := secp256k1.SignatureFromBytes(signature)
	if err != nil {
		return false, err
	}

	// Parse the public key
	pubKey, err := secp256k1.ParsePubKey(pk.data)
	if err != nil {
		return false, err
	}

	return sig.Verify(messageHash, pubKey), nil
} 